src/Lib.hs

```diff
@@ -43,76 +48,183 @@ startPetSelection user goldRemaining = do
+  let p2HealthRemaining = getHealth (petHealthRemaining p2) - getAttack (petAttack p1)
+
+  -- create new pets with new health
+  let p1' = p1 { petHealthRemaining = (Health p1HealthRemaining) }
+  let p2' = p2 { petHealthRemaining = (Health p2HealthRemaining) }
+  
+  -- check results
+  if ((getHealth (petHealthRemaining p1') < 0) || (getHealth (petHealthRemaining p2') < 0))
+    then (p1', p2')  -- one or both pets are dead
+    else battlePets p1' p2'  -- not dead battle again
+
+-- keep asking
+keepAskingWhere :: Read a => String -> (a -> Bool) -> IO a
+keepAskingWhere s p = do
+  putStr s
+  hFlush stdout
+  mresult <- readMaybe <$> getLine
+  case mresult of
+    Nothing -> keepAskingWhere s p
+    Just a -> if p a then pure a else keepAskingWhere s p
+
+
+keepAsking :: Read a => String -> IO a
+keepAsking s = keepAskingWhere s (const True)
+
+
+betweenInclusive :: Int -> Int -> Int -> Bool
+betweenInclusive a b x = a <= x && x <= b
 
 
 --
 -- ATTRIBUTES
 --
-newtype Name = Name {getName::String} deriving (Show, IsString)
-newtype Attack = Attack {getAttack::Int} deriving Show
-newtype Health = Health {getHealth::Int} deriving Show
+newtype Id = Id {getId::Int} deriving (Num, Ord, Eq)
+newtype Name = Name {getName::String} deriving (IsString, Eq)
+newtype Emoji = Emoji {getEmoji::String} deriving (IsString, Eq)
+newtype Attack = Attack {getAttack::Int} deriving (Num, Ord, Eq)
+newtype Health = Health {getHealth::Int} deriving (Num, Ord, Eq)
 newtype Cost = Cost {getCost::Int} deriving (Num, Ord, Eq)
 
-
 class Attribute a where
-    valid :: a -> Bool
+  valid :: a -> Bool
+
+instance Attribute Id where
+  valid (Id x) = x > 0
 
 instance Attribute Name where
-    valid (Name x) = x /= ""
+  valid (Name x) = x /= ""
 
 instance Attribute Attack where
-    valid (Attack x) = x > 0
+  valid (Attack x) = x > 0
 
 instance Attribute Health where
-    valid (Health x) = x > 0
+  valid (Health x) = x > 0
 
 instance Attribute Cost where
-    valid (Cost x) = x >= 0
-
-instance Show Cost where
-    show (Cost x) = show x
+  valid (Cost x) = x >= 0
 
 
 --
 -- PETS
 --
 data Pet = Pet
-  { petName :: Name
+  { petId :: Id
+  , petName :: Name
+  , petEmoji :: Emoji
   , petAttack :: Attack
   , petHealth :: Health
+  , petHealthRemaining :: Health
   , petCost :: Cost
   }
--- TODO how to do this? is this the best way to do this? is this necessary?
--- data AnimalType = Turtle | Ant | Penguin | etc...
 
 
-instance Show Pet where
-  show (Pet name attack health cost) = (getName name) ++ " $" ++ show (getCost cost) ++ " (A: " ++ show (getAttack attack) ++ ", H: " ++ show (getHealth health) ++ ")"
+instance Eq Pet where
+   (Pet id name emoji attack healthRemaining health cost) == (Pet id' name' emoji' attack' healthRemaining' health' cost') = id == id'
+
+
+mkPet :: Name -> Id -> Emoji -> Attack -> Health -> Cost -> Pet
+mkPet name id emoji attack health cost = Pet
```

Comment:
you are shadowing the function `id` here, ideally you would choose another name. I would probably consider using `pid`